src/Accelerators/NNPA/Transform/ZHigh/ZHighLayoutPropagation.td

```diff
@@ -31,185 +31,221 @@ include "src/Accelerators/NNPA/Dialect/ZHigh/ZHigh.td"
 // Pattern-Match and Rewrite
 //===----------------------------------------------------------------------===//
 
-// Check values of an ArrayAttr.
-def IsPermToNHWC:
-  Constraint<CPred<"$_self.getValue().size() == 4 && "
-                   "($_self.getValue()[0]).cast<IntegerAttr>().getInt() == 0 && "
-                   "($_self.getValue()[1]).cast<IntegerAttr>().getInt() == 2 && "
-                   "($_self.getValue()[2]).cast<IntegerAttr>().getInt() == 3 && "
-                   "($_self.getValue()[3]).cast<IntegerAttr>().getInt() == 1">,
-                 "Permutation is [0, 2, 3, 1]">;
-
-// Check values of an ArrayAttr.
-def IsPermToNCHW:
-  Constraint<CPred<"$_self.getValue().size() == 4 && "
-                   "($_self.getValue()[0]).cast<IntegerAttr>().getInt() == 0 && "
-                   "($_self.getValue()[1]).cast<IntegerAttr>().getInt() == 3 && "
-                   "($_self.getValue()[2]).cast<IntegerAttr>().getInt() == 1 && "
-                   "($_self.getValue()[3]).cast<IntegerAttr>().getInt() == 2">,
-                 "Permutation is [0, 3, 1, 2]">;
-
 def GetLayout : NativeCodeCall<
   "convertZTensorDataLayoutToStringAttr($_builder, ::onnx_mlir::zhigh::getZTensorLayout($0.getType()))"
 >;
 
-def IsNHWC: Constraint<
-  CPred<"getZTensorLayout($0.getType()) == ::onnx_mlir::zhigh::ZTensorEncodingAttr::DataLayout::NHWC">,
-  "Check if a value is NHWC"
+def LayoutIs: Constraint<
+  CPred<"convertZTensorDataLayoutToStringAttr($_builder, ::onnx_mlir::zhigh::getZTensorLayout($_self.getType())) == $0">,
+  "Check if the given value's layout is the same as the given layout"
+>;
+
+def LayoutIsNot: Constraint<
+  CPred<"convertZTensorDataLayoutToStringAttr($_builder, ::onnx_mlir::zhigh::getZTensorLayout($_self.getType())) != $0">,
+  "Check if the given value's layout is different from the given layout"
+>;
+
+//===----------------------------------------------------------------------===//
+// Layout propagation for unary element-wise operations
+//
+// Unary operations, such as Relu and Tanh, do not depend on a specific layout.
+// Thus, it is possible to use the layout of the previous zTensor. In other
+// words, propagate the existing layout down to the unary operation, so that
+// other rules can be applied to remove unstick-stick pairs.For example,
+// This sequence of operations:
+//   zhigh.Conv -> Unstick (NHWC) -> Stick (4D) -> zhigh.Relu
+// will become:
+//   zhigh.Conv -> zhigh.Relu -> Unstick (NHWC) -> Stick (4D)
+// then, canonicalization of unstick/stick will remove the unstick-stick pair.
+//
+//===----------------------------------------------------------------------===//
+
+// Exp
```

Comment:
Quick question: is it possible to make a rule that 1) detect if the op has the "layout oblivious" property and if it does, that the stick/unstick ops are moved around? It is true that the set of ops is more limited that the ONNX standard, so there are not that many, but it might still be a nice feature to have.

Reply:
Yes, it's possible, just use tablegen for easy tracking since there are not many zHigh ops as you know. If you prefer I can change it to writing C++ instead of tablegen. Perhaps, we would have the same thing for CPU layout in the future.